generative_data_prep/utils/utils.py

```diff
@@ -255,6 +260,92 @@ def execute_and_return_stdout(command):
     return result
 
 
+def _calculate_sha256(file_path: str):
+    sha256_hash = hashlib.sha256()
+    with open(file_path, "rb") as file:
+        for byte_block in iter(lambda: file.read(4096), b""):
+            sha256_hash.update(byte_block)
+    return sha256_hash.hexdigest()
+
+
+def _get_walk_files_to_hash(dir: str, filter: Optional[str] = None):
+    files_to_hash = []
+    for foldername, _, filenames in os.walk(dir):
+        if filter is not None and filter in foldername.split("/"):
+            continue
+        relative_foldername = os.path.relpath(foldername, dir)
+        if relative_foldername == ".":
+            relative_foldername = ""
+        relative_foldername = relative_foldername.replace(os.path.sep, "_")
+        if relative_foldername != "":
+            relative_foldername += "_"
+        hash_file_name = [
+            (
+                os.path.join(foldername, filename),
+                relative_foldername + filename.split(".")[0] + ".txt",
+            )
+            for filename in filenames
+        ]
+
+        files_to_hash.extend(hash_file_name)
+    return files_to_hash
+
+
+def validate_sha256(output_dir: str):
+    """Validates the current sha256 directory within the output directory.
+
+    Args:
+        output_dir (str)
+
+    Returns:
+        Boolean value: False for the files have been modified, and True for all
+        files have not been modified
+    """
+    files_to_hash = _get_walk_files_to_hash(output_dir, "sha256")
+    sha_info_file = os.path.join(output_dir, "sha256", "files_metadata.json")
+    with open(sha_info_file, "r") as output_file:
+        file_info_dict = json.load(output_file)
+    for file, hash_file_name in files_to_hash:
+        if hash_file_name not in file_info_dict:
+            return False
+        current_modified_time = os.path.getmtime(file)
+        current_size = os.path.getsize(file)
+        if current_size != file_info_dict[hash_file_name]["size"] or not math.isclose(
+            current_modified_time, file_info_dict[hash_file_name]["modified_time"]
+        ):
+            file_hash = file_info_dict[hash_file_name]["sha256"]
+            current_file_hash = _calculate_sha256(file)
+            if file_hash != current_file_hash:
+                return False
+    return True
```

Comment:
nice! we will not call this during training though, right? My only worry is that if they want to customize the dataset by adding files or something it would fail out?

Reply:
hmmmm I guess we could have some way where if a user wants to add some new files to the dataset we can have a function that can take in the new path and generate the sha256 and add it to the hash file? Does that sound like a good idea to you? And I guess this is more of a failsafe, so if it passes the inital metadata then it will not scrub the dataset by looking into each hdf5 file, but if it fails the metadata check then it will do the current runtime checks we do where it looks into each file!

Reply:
Not sure I quite understand, but I trust.

Another option is to just fail out during training and allow the user to over-ride the checking with a flag like --do_not_verify_input_data if they customize it?

Reply:
ah I guess its kind of like this

use sha256 to verify that none of the files have been modified:

if the files have been modified then we do stage 1 (going through each hdf5 files and extracting the necessary data components which should take a while) https://github.sambanovasystems.com/SambaNova/software/blob/151953ccb83dadb5ad70986d2f054a404dd87584/sambaflow/apps/nlp/transformers_on_rdu/transformers_hook_v2.py#L168-L171

if the files have not been modified then we do stage 2 (this PR where we just check using the metadata file)

but you're right in that we should have some type of override check, I guess for sambastudio what I imagined is that they could do stage 2 in SS suite before even alloc resources so you dont have to wait for RDUs to open up, and if it passes then we can pass a flag to the modelbox saying there is no need to do stage 1!

Reply:
This makes perfect sense now! 99% of the time we should be able to work with stage 2! 

Make sure that you account for the fact that users can copy in new files, not just modify the existing files